cdc/sink/codec/canal/canal_json_encoder.go

```diff
@@ -24,133 +24,162 @@ import (
+				mysqlTypeMap[column.Name] = mysqlType
+			}
 
-	if len(rowData.BeforeColumns) > 0 {
-		oldData = make(map[string]interface{}, len(rowData.BeforeColumns))
-		for i := range rowData.BeforeColumns {
-			if !rowData.BeforeColumns[i].GetIsNull() {
-				oldData[rowData.BeforeColumns[i].Name] = rowData.BeforeColumns[i].Value
+			if column.Value == nil {
+				oldData[column.Name] = nil
 			} else {
-				oldData[rowData.BeforeColumns[i].Name] = nil
+				oldData[column.Name] = value
 			}
 		}
 	}
 
-	if len(rowData.AfterColumns) > 0 {
-		data = make(map[string]interface{}, len(rowData.AfterColumns))
-		for i := range rowData.AfterColumns {
-			if !rowData.AfterColumns[i].GetIsNull() {
-				data[rowData.AfterColumns[i].Name] = rowData.AfterColumns[i].Value
+	if len(e.Columns) > 0 {
+		data = make(map[string]interface{}, len(e.Columns))
+	}
+	for _, column := range e.Columns {
+		if column != nil {
+			mysqlType := getMySQLType(column)
+			javaType, err := getJavaSQLType(column, mysqlType)
+			if err != nil {
+				return cerror.WrapError(cerror.ErrCanalEncodeFailed, err)
+			}
+			value, err := c.builder.formatValue(column.Value, javaType)
+			if err != nil {
+				return cerror.WrapError(cerror.ErrCanalEncodeFailed, err)
+			}
+			if !isDelete {
+				sqlTypeMap[column.Name] = int32(javaType)
+				mysqlTypeMap[column.Name] = mysqlType
+			}
+			if column.Value == nil {
+				data[column.Name] = nil
 			} else {
-				data[rowData.AfterColumns[i].Name] = nil
+				data[column.Name] = value
 			}
 		}
 	}
 
-	msg := &canalJSONMessage{
-		ID:            0, // ignored by both Canal Adapter and Flink
-		Schema:        header.SchemaName,
-		Table:         header.TableName,
-		PKNames:       pkNames,
-		IsDDL:         false,
-		EventType:     header.GetEventType().String(),
-		ExecutionTime: header.ExecuteTime,
-		BuildTime:     time.Now().UnixNano() / 1e6, // ignored by both Canal Adapter and Flink
-		Query:         "",
-		SQLType:       sqlType,
-		MySQLType:     mysqlType,
-		Data:          make([]map[string]interface{}, 0),
-		Old:           nil,
-		tikvTs:        e.CommitTs,
+	var baseMessage *canalJSONMessage
+	if !c.enableTiDBExtension {
+		baseMessage = c.messageHolder.(*canalJSONMessage)
+	} else {
+		baseMessage = c.messageHolder.(*canalJSONMessageWithTiDBExtension).canalJSONMessage
 	}
 
+	baseMessage.ID = 0 // ignored by both Canal Adapter and Flink
+	baseMessage.Schema = e.Table.Schema
+	baseMessage.Table = e.Table.Table
+	baseMessage.PKNames = e.PrimaryKeyColumnNames()
+	baseMessage.IsDDL = false
+	baseMessage.EventType = eventTypeString(e)
+	baseMessage.ExecutionTime = convertToCanalTs(e.CommitTs)
+	baseMessage.BuildTime = time.Now().UnixNano() / 1e6 // ignored by both Canal Adapter and Flink
+	baseMessage.Query = ""
+	baseMessage.SQLType = sqlTypeMap
+	baseMessage.MySQLType = mysqlTypeMap
+	baseMessage.Old = nil
+	baseMessage.tikvTs = e.CommitTs
+
 	if e.IsDelete() {
-		msg.Data = append(msg.Data, oldData)
+		baseMessage.Data[0] = oldData
```

Comment:
What if we make `Data` field just a `map[string]interface{}` since it only contains one item?

Reply:
it's represented as `an array with only one item` officially.